File: script/tools/enable_auth_websockets.py
#!/usr/bin/env python3

import json
import shutil
import sys
from datetime import datetime
from pathlib import Path


def usage():
    print("Usage: python3 enable_auth_websockets.py <auth-dir> [--dry-run] [--no-backup]")


def main():
    if len(sys.argv) < 2:
        usage()
        return 1

    auth_dir = Path(sys.argv[1]).expanduser().resolve()
    dry_run = "--dry-run" in sys.argv
    no_backup = "--no-backup" in sys.argv
    stamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    stats = {}

    for file in auth_dir.rglob("*.json"):
        try:
            data = json.loads(file.read_text(encoding="utf-8"))
        except Exception:
            stats["skipped"] = stats.get("skipped", 0) + 1
            print(f"skipped: {file} (invalid_json)")
            continue

        if not isinstance(data, dict):
            stats["skipped"] = stats.get("skipped", 0) + 1
            print(f"skipped: {file} (json_root_not_object)")
            continue

        if data.get("websockets") is True:
            stats["unchanged"] = stats.get("unchanged", 0) + 1
            print(f"unchanged: {file}")
            continue

        data["websockets"] = True

        if dry_run:
            stats["would_update"] = stats.get("would_update", 0) + 1
            print(f"would_update: {file}")
            continue

        if not no_backup:
            shutil.copy2(file, f"{file}.bak-{stamp}")

        file.write_text(json.dumps(data, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

        stats["updated"] = stats.get("updated", 0) + 1
        print(f"updated: {file}")

    print("summary:", stats)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: script/tools/enable-auth-websockets.mjs
#!/usr/bin/env node

import { copyFile, readdir, readFile, rename, writeFile } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';

const args = process.argv.slice(2);
const dryRun = args.includes('--dry-run');
const noBackup = args.includes('--no-backup');
const authDirArg = args.find((arg) => !arg.startsWith('--'));

if (!authDirArg) {
  console.error('Usage: node enable-auth-websockets.mjs <auth-dir> [--dry-run] [--no-backup]');
  process.exit(1);
}

const authDir = resolve(authDirArg);
const stamp = new Date().toISOString().replace(/[:.]/g, '-');

async function walk(dir) {
  const entries = await readdir(dir, { withFileTypes: true });
  const files = [];

  for (const entry of entries) {
    const fullPath = join(dir, entry.name);

    if (entry.isDirectory()) {
      files.push(...(await walk(fullPath)));
      continue;
    }

    if (entry.isFile() && entry.name.endsWith('.json')) {
      files.push(fullPath);
    }
  }

  return files;
}

async function updateJson(filePath) {
  let raw;

  try {
    raw = await readFile(filePath, 'utf8');
  } catch (error) {
    return { filePath, status: 'read_failed', reason: error.message };
  }

  let data;

  try {
    data = JSON.parse(raw);
  } catch {
    return { filePath, status: 'skipped', reason: 'invalid_json' };
  }

  if (!data || Array.isArray(data) || typeof data !== 'object') {
    return { filePath, status: 'skipped', reason: 'json_root_not_object' };
  }

  if (data.websockets === true) {
    return { filePath, status: 'unchanged' };
  }

  data.websockets = true;

  if (dryRun) {
    return { filePath, status: 'would_update' };
  }

  if (!noBackup) {
    await copyFile(filePath, `${filePath}.bak-${stamp}`);
  }

  const tmpPath = join(dirname(filePath), `.enable-auth-websockets-${process.pid}-${Math.random().toString(16).slice(2)}.tmp`);
  await writeFile(tmpPath, `${JSON.stringify(data, null, 2)}\n`, 'utf8');
  await rename(tmpPath, filePath);

  return { filePath, status: 'updated' };
}

const files = await walk(authDir);
const results = [];

for (const filePath of files) {
  results.push(await updateJson(filePath));
}

const summary = results.reduce((acc, item) => {
  acc[item.status] = (acc[item.status] || 0) + 1;
  return acc;
}, {});

for (const item of results) {
  const suffix = item.reason ? ` (${item.reason})` : '';
  console.log(`${item.status}: ${item.filePath}${suffix}`);
}

console.log('summary:', JSON.stringify(summary));
